File: 백준/Gold/13913. 숨바꼭질 4/숨바꼭질 4.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.Stack
import java.util.StringTokenizer

fun main(){
    val br = BufferedReader(InputStreamReader(System.`in`))
    val q = ArrayDeque<Int>()
    val token = StringTokenizer(br.readLine())
    val n = token.nextToken().toInt()
    val k = token.nextToken().toInt()
    val d = Array<Int>(2*maxOf(n,k,10)){
        -1
    }
    val from = Array<Int>(2*maxOf(n,k,10)){
        0
    }
    val MAX = 2*k
    q.add(n)
    d[n] = 0
    while(!q.isEmpty()){
        val now = q.removeFirst()
        if(now + 1 < MAX){
            if(d[now+1] == -1) {
                q.add(now + 1)
                d[now + 1] = d[now] + 1
                from[now + 1] = now
            }
        }

        if(now - 1 >= 0){
            if(d[now-1] == -1) {
                q.add(now - 1)
                d[now - 1] = d[now] + 1
                from[now-1] = now
            }
        }

        if(now * 2 < MAX){
            if(d[now*2] == -1) {
                q.add(now*2)
                d[now*2] = d[now] + 1
                from[now*2] = now
            }
        }
    }

    val stc = Stack<Int>()
    println(d[k])
    var temp:Int = k
    stc.push(k)
    repeat(d[k]){
        temp = from[temp]
        stc.push(temp)
    }
    while(!stc.isEmpty()){
        print("${stc.pop()} ")
    }
}

File: 백준/Gold/13913. 숨바꼭질 4/숨바꼭질 4.cc
#include<iostream>
#include<queue>

using namespace std;

const int MAX = 200000;
bool visited[MAX + 1];
int d[MAX + 1];
int from[MAX + 1];

void print(int n, int m) {
	if (n != m) {
		print(n, from[m]);
	}

	cout << m << ' ';
}
int main() {
	int n, k;
	cin >> n >> k;
	
	visited[n] = true;
	queue<int> q;
	
	q.push(n);

	while (!q.empty()) {
		int now = q.front();
		q.pop();

		if (now - 1 >= 0) {
			if (!visited[now - 1]) {
				q.push(now - 1);
				visited[now - 1] = true;
				d[now - 1] = d[now] + 1;
				from[now - 1] = now;
			}
		}

		if (now + 1 < MAX) {
			if (!visited[now + 1]) {
				q.push(now + 1);
				visited[now + 1] = true;
				d[now + 1] = d[now] + 1;
				from[now + 1] = now;
			}
		}

		if (now * 2 < MAX) {
			if (!visited[now * 2]) {
				q.push(now * 2);
				visited[now * 2] = true;
				d[now * 2] = d[now] + 1;
				from[now * 2] = now;
			}
		}
	}

	cout << d[k] << "\n";
	print(n, k);

	cout << "\n";

	return 0;
}